contrib/Joshua/scripts/correctnessTest.sh

```diff
@@ -1,10 +1,176 @@
+# Called by the Joshua testing framework.
+# See contrib/TestHarness2/README.md for detailed documentation.
 
-# export RARE_PRIORITY=20
-python3 -m test_harness.app -s ${JOSHUA_SEED} --old-binaries-path ${OLDBINDIR}
+# Cleanup function - preserve logs on failure if archival is enabled
+cleanup() {
+    echo "--- correctnessTest.sh cleanup starting ---" >&2
+    
+    local preserve_files=false
+    local test_failed=false
+    
+    # Check if test failed by parsing XML output
+    if [ -f "${PYTHON_APP_STDOUT_FILE}" ] && grep -q 'Ok="0"' "${PYTHON_APP_STDOUT_FILE}"; then
+        test_failed=true
+    fi
+    
+    # Preserve files if:
+    # 1. Always preserve flag is set, OR
+    # 2. (Python crashed OR test failed) AND archival is enabled
+    if [ "${TH_PRESERVE_TEMP_DIRS_ON_EXIT}" = "true" ] || \
+       ( ([ "${PYTHON_EXIT_CODE}" -ne "0" ] || [ "${test_failed}" = "true" ]) && [ "${TH_ARCHIVE_LOGS_ON_FAILURE}" = "true" ] ); then
+        preserve_files=true
+    fi
+    
+    if [ "${preserve_files}" = "true" ]; then
+        echo "Preserving test artifacts in: ${TOP_LEVEL_OUTPUT_DIR}" >&2
+        if [ "${test_failed}" = "true" ] || [ "${PYTHON_EXIT_CODE}" -ne "0" ]; then
+            echo "Test failed - logs preserved for debugging. Use 'kubectl cp' to copy from pod." >&2
+        fi
+    else
+        echo "Cleaning up test artifacts: ${TOP_LEVEL_OUTPUT_DIR}" >&2
+        rm -rf "${TOP_LEVEL_OUTPUT_DIR}"
+    fi
+}
+
+trap cleanup EXIT
+
+# Set defaults for key environment variables
+export ASAN_OPTIONS="${ASAN_OPTIONS:-detect_leaks=0}"
+export TH_DISABLE_ROCKSDB_CHECK="${TH_DISABLE_ROCKSDB_CHECK:-false}"
+
+# Setup unique output directory
+TH_OUTPUT_BASE_DIR="${TH_OUTPUT_DIR:-${DIAG_LOG_DIR:-/tmp}}"
+
+# Use ensemble ID if available, otherwise fall back to joshua seed
+if [ -n "${JOSHUA_ENSEMBLE_ID}" ]; then
+    UNIQUE_RUN_SUFFIX="${JOSHUA_ENSEMBLE_ID}"
+    echo "Using ensemble ID for directory name: ${UNIQUE_RUN_SUFFIX}" >&2
+else
+    # Try to extract ensemble ID from working directory (like joshua_logtool.py does)
+    EXTRACTED_ENSEMBLE_ID=$(echo "${PWD}" | grep -o 'ensembles/[0-9A-Za-z._-]*' | cut -d'/' -f2)
+    if [ -n "${EXTRACTED_ENSEMBLE_ID}" ]; then
+        UNIQUE_RUN_SUFFIX="${EXTRACTED_ENSEMBLE_ID}"
+        echo "Extracted ensemble ID from working directory: ${UNIQUE_RUN_SUFFIX}" >&2
+    else
+        UNIQUE_RUN_SUFFIX="${JOSHUA_SEED}"
+        echo "Using joshua seed for directory name: ${UNIQUE_RUN_SUFFIX}" >&2
+    fi
+fi
+
+TOP_LEVEL_OUTPUT_DIR="${TH_OUTPUT_BASE_DIR}/th_run_${UNIQUE_RUN_SUFFIX}"
+
+APP_RUN_TEMP_DIR="${TOP_LEVEL_OUTPUT_DIR}"
+
+# Create directories
+mkdir -p "${APP_RUN_TEMP_DIR}"
+if [ ! -d "${APP_RUN_TEMP_DIR}" ]; then
+    echo "FATAL: Failed to create required directories" >&2
+    exit 1
+fi
+
+# Set permissions
+chmod 777 "${TOP_LEVEL_OUTPUT_DIR}"
+
+# Setup Python path for in-tree modules
+SCRIPT_DIR=$(dirname "$0")
+REPO_ROOT=$(cd "$SCRIPT_DIR/../../.." && pwd)
```

Comment:
On Joshua pod, we don't have FDB repo. Is this for okteto environment?

Reply:
I can imagine why I wrote it that way but can't understand why it ever worked -- even adding debugging it shouldn't have worked (but does). Fixing.  Good one.